mlflow/data/tensorflow_dataset.py

```diff
@@ -25,22 +29,47 @@ class TensorflowDataset(Dataset, PyFuncConvertibleDatasetMixin):
 
     def __init__(
         self,
-        data,
+        features,
         source: DatasetSource,
         targets=None,
         name: Optional[str] = None,
         digest: Optional[str] = None,
     ):
         """
-        :param data: A TensorFlow dataset or tensor.
+        :param features: A TensorFlow dataset or tensor of features.
         :param source: The source of the TensorFlow dataset.
-        :param targets: A TensorFlow dataset or tensor containing dataset targets. Optional
+        :param targets: A TensorFlow dataset or tensor of targets. Optional.
         :param name: The name of the dataset. E.g. "wiki_train". If unspecified, a name is
                      automatically generated.
         :param digest: The digest (hash, fingerprint) of the dataset. If unspecified, a digest
                        is automatically computed.
         """
-        self._data = data
+        if not isinstance(features, tf.data.Dataset) and not tf.is_tensor(features):
+            raise MlflowException(
+                f"'features' must be an instance of tf.data.Dataset or a TensorFlow Tensor."
+                f" Found: {type(features)}.",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        if tf.is_tensor(features) and targets is not None and not tf.is_tensor(targets):
+            raise MlflowException(
+                f"If 'features' is a TensorFlow Tensor, then 'targets' must also be a TensorFlow"
+                f" Tensor. Found: {type(targets)}.",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        if (
+            isinstance(features, tf.data.Dataset)
+            and targets is not None
+            and not isinstance(targets, tf.data.Dataset)
+        ):
+            raise MlflowException(
+                "If 'features' is an instance of tf.data.Dataset, then 'targets' must also be an"
+                f" instance of tf.data.Dataset. Found: {type(targets)}.",
+                INVALID_PARAMETER_VALUE,
+            )
```

Comment:
Our TF dataset doesn't currently handle the case where features and targets are different types. Added some validation logic here for that.